pkg/agent/client/endpoint_resolver.go

```diff
@@ -0,0 +1,292 @@
+// (whether it's kube-proxy or AntreaProxy). A new Endpoint is resolved every time the Service's
+// Spec or the Ednpoints' Subsets are updated, and registered listeners are notified. While this
+// EndpointResolver is somewhat generic, at the moment it is only meant to be used for the Antrea
+// Service.
+type EndpointResolver struct {
+	// name is the name of the controller in charge of Endpoint resolution.
+	name        string
+	namespace   string
+	serviceName string
+	servicePort int32
+	// informerFactory is stored here so it can be started in the Run() method.
+	informerFactory informers.SharedInformerFactory
+	// serviceLister is used to retrieve the Service when selecting an Endpoint.
+	serviceLister       corev1listers.ServiceLister
+	serviceListerSynced cache.InformerSynced
+	// endpointLister is used to retrieve the Endpoints for the Service during Endpoint selection.
+	endpointsLister       corev1listers.EndpointsLister
+	endpointsListerSynced cache.InformerSynced
+	queue                 workqueue.RateLimitingInterface
+	// listeners need to implement the Listerner interface and will get notified when the
+	// current Endpoint URL changes.
+	listeners   []Listener
+	endpointURL atomic.Pointer[url.URL]
+}
+
+func NewEndpointResolver(kubeClient kubernetes.Interface, namespace, serviceName string, servicePort int32) *EndpointResolver {
+	key := namespace + "/" + serviceName
+	controllerName := fmt.Sprintf("ServiceEndpointResolver-%s", key)
+
+	// We only need a specific Service and corresponding Endpoints resource, so we create our
+	// own informer factory, and we filter by namespace and name.
+	informerFactory := informers.NewSharedInformerFactoryWithOptions(kubeClient, informerDefaultResync, informers.WithNamespace(namespace), informers.WithTweakListOptions(func(listOptions *metav1.ListOptions) {
+		listOptions.FieldSelector = fields.OneTermEqualSelector("metadata.name", serviceName).String()
+	}))
+	serviceInformer := informerFactory.Core().V1().Services()
+	endpointsInformer := informerFactory.Core().V1().Endpoints()
+
+	resolver := &EndpointResolver{
+		name:                  controllerName,
+		namespace:             namespace,
+		serviceName:           serviceName,
+		servicePort:           servicePort,
+		informerFactory:       informerFactory,
+		serviceLister:         serviceInformer.Lister(),
+		serviceListerSynced:   serviceInformer.Informer().HasSynced,
+		endpointsLister:       endpointsInformer.Lister(),
+		endpointsListerSynced: endpointsInformer.Informer().HasSynced,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), controllerName),
+	}
+
+	serviceInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		// FilterFunc ignores all Service events which do not relate to the named Service.
+		// It should be redudant given the filtering that we already do at the informer level.
```

Comment:
s/redudant/redundant, but do we still need to keep the filter?

Reply:
I feel like it's ok to keep it. I saw the same pattern in `ConfigMapCAController`, even though I agree it is not needed.